### PR TITLE
Support modifying a TextRegionAccess

### PR DESCRIPTION
**Pull request by meysholdt**
This is 
- the basis for https://github.com/eclipse/xtext-core/issues/387
- the basis for document modification with subsequent formatting.

### CODE REVIEW

**Comment by meysholdt** [comment]
Reviews are very welcome.

File: org.eclipse.xtext.tests/src/org/eclipse/xtext/formatting2/regionaccess/internal/RegionAccessDiffTest.xtend [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+
+	@Test def void testDeleteOne() {
+		val access = '''
+			4/*1*/prefix/*2*/foo
+		'''.toTextRegionAccess
+		access.modify [
+			val extension ext = access.extensions
+			val rootObj = access.regionForRootEObject.semanticElement as PrefixedUnassigned
+			val prefix = rootObj.delegate.regionFor.keyword("prefix")
+			remove(prefix, prefix)
```
Comment (svenefftinge) [inline org.eclipse.xtext.tests/src/org/eclipse/xtext/formatting2/regionaccess/internal/RegionAccessDiffTest.xtend]: remove seems to be inclusive, while replace is exclusive

File: org.eclipse.xtext.tests/src/org/eclipse/xtext/formatting2/regionaccess/internal/RegionAccessCommentsTest.xtend [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+import org.eclipse.xtext.testing.XtextRunner
+import org.junit.Test
+import org.junit.runner.RunWith
+
+/**
+ * @author Moritz Eysholdt - Initial contribution and API
+ */
+@InjectWith(RegionAccessTestLanguageInjectorProvider)
+@RunWith(XtextRunner)
+class RegionAccessCommentsTest {
```
Comment (cdietrich) [inline org.eclipse.xtext.tests/src/org/eclipse/xtext/formatting2/regionaccess/internal/RegionAccessCommentsTest.xtend]: Tests fail on Windows

Comment (cdietrich) [inline org.eclipse.xtext.tests/src/org/eclipse/xtext/formatting2/regionaccess/internal/RegionAccessCommentsTest.xtend]:
```
org.junit.ComparisonFailure: expected:<..._COMMENT'
        "\[n"       Whitespace:TerminalRule'WS'
        "/*a*/"    Comment:TerminalRule'ML_COMMENT'
   12   "\n"       Whitespace:TerminalRule'WS'
        B ValueList'[a]' Root
12  1    S "a"        ValueList:name+=ID
        E ValueList'[a]' Root
13  0   H
------------ diff 1 ------------
 0    H "/*h*/"    Comment:TerminalRule'ML_COMMENT'
        "\n"       Whitespace:TerminalRule'WS'
        "/*8*/"    Comment:TerminalRule'ML_COMMENT'
   12   "\n"       Whitespace:TerminalRule'WS'
12  1 S "8"        Root:'8'
13    H " "        Whitespace:TerminalRule'WS'
        "//8\n"    Comment:TerminalRule'SL_COMMENT'
        "/*a*/"    Comment:TerminalRule'ML_COMMENT'
   11   "\n"  ]     Whitespace:Term...> but was:<..._COMMENT'
        "\[r\n"     Whitespace:TerminalRule'WS'
        "/*a*/"    Comment:TerminalRule'ML_COMMENT'
   14   "\r\n"     Whitespace:TerminalRule'WS'
        B ValueList'[a]' Root
14  1    S "a"        ValueList:name+=ID
        E ValueList'[a]' Root
15  0   H
------------ diff 1 ------------
 0    H "/*h*/"    Comment:TerminalRule'ML_COMMENT'
        "\r\n"     Whitespace:TerminalRule'WS'
        "/*8*/"    Comment:TerminalRule'ML_COMMENT'
   14   "\r\n"     Whitespace:TerminalRule'WS'
14  1 S "8"        Root:'8'
15    H " "        Whitespace:TerminalRule'WS'
        "//8\r\n"  Comment:TerminalRule'SL_COMMENT'
        "/*a*/"    Comment:TerminalRule'ML_COMMENT'
   13   "\r\n"]     Whitespace:Term...>
	at org.junit.Assert.assertEquals(Assert.java:115)
	at org.junit.Assert.assertEquals(Assert.java:144)
```

Comment (kittaakos) [inline org.eclipse.xtext.tests/src/org/eclipse/xtext/formatting2/regionaccess/internal/RegionAccessCommentsTest.xtend]: That was the fix each time, I have stumbled upon such a Windows line ending issue. And I have to create a utility class for each project. I know it is kind of off-topic, but wouldn't it make sense to introduce such a helper in Xtext?

Comment (meysholdt) [inline org.eclipse.xtext.tests/src/org/eclipse/xtext/formatting2/regionaccess/internal/RegionAccessCommentsTest.xtend]: good pointer. yes. Somethings like this: https://github.com/eclipse/xtext-core/pull/416/commits/7fb40853079f41ee35136b56a275d530df60b021#diff-d530da1e414f60b23880c98e1a61de59R105 seems helpful.

**Comment by meysholdt** [comment]
Thank you for the feedback - it's all resolved in the latest patch set. I intend to merge tomorrow morning unless further issues pop up.

**Comment by cdietrich** [comment]
still have failing tests
```
org.junit.ComparisonFailure: expected:<...COMMENT'
        "\[n"       Whitespace:TerminalRule'WS'
        "/*a*/"    Comment:TerminalRule'ML_COMMENT'
   12   "\n"       Whitespace:TerminalRule'WS'
        B ValueList'[a]' Root
12  1    S "a"        ValueList:name+=ID
        E ValueList'[a]' Root
13  0   H
------------ diff 1 ------------
 0    H "/*h*/"    Comment:TerminalRule'ML_COMMENT'
        "\n"       Whitespace:TerminalRule'WS'
        "/*8*/"    Comment:TerminalRule'ML_COMMENT'
   12   "\n"       Whitespace:TerminalRule'WS'
12  1 S "8"        Root:'8'
13    H " "        Whitespace:TerminalRule'WS'
        "//8\n"    Comment:TerminalRule'SL_COMMENT'
        "/*a*/"    Comment:TerminalRule'ML_COMMENT'
   11   "\n"  ]     Whitespace:Term...> but was:<...COMMENT'
        "\[r\n"     Whitespace:TerminalRule'WS'
        "/*a*/"    Comment:TerminalRule'ML_COMMENT'
   14   "\r\n"     Whitespace:TerminalRule'WS'
        B ValueList'[a]' Root
14  1    S "a"        ValueList:name+=ID
        E ValueList'[a]' Root
15  0   H
------------ diff 1 ------------
 0    H "/*h*/"    Comment:TerminalRule'ML_COMMENT'
        "\r\n"     Whitespace:TerminalRule'WS'
        "/*8*/"    Comment:TerminalRule'ML_COMMENT'
   14   "\r\n"     Whitespace:TerminalRule'WS'
14  1 S "8"        Root:'8'
15    H " "        Whitespace:TerminalRule'WS'
        "//8\r\n"  Comment:TerminalRule'SL_COMMENT'
        "/*a*/"    Comment:TerminalRule'ML_COMMENT'
   13   "\r\n"]     Whitespace:Term...>
	at org.junit.Assert.assertEquals(Assert.java:115)
	at org.junit.Assert.assertEquals(Assert.java:144)

```

**Comment by meysholdt** [comment]
is it because of lineSeparators?

**Comment by cdietrich** [comment]
seems so. change test to

```
@Test def void testDeleteFirstRegion1() {
		System.setProperty("line.separator","\r\n")
``` and you can see the problem on mac/linux as well

**Comment by meysholdt** [comment]
found it. fixed.

**Comment by cdietrich** [comment]
looks good now

**Pull request merged**